tests/__init__.py

```diff
@@ -1,74 +1,185 @@
+
         # Assert that the ingredients returned by the ingredient_groups() function
         # are the same as the ingredients return by the ingredients() function.
         grouped = []
-        for group in self.harvester_class.ingredient_groups():
+        for group in actual.ingredient_groups():
             grouped.extend(group.ingredients)
 
-        self.assertEqual(sorted(self.harvester_class.ingredients()), sorted(grouped))
-
-    def test_multiple_instructions(self):
-        # Assert that the instructions_list() method returns more than one item;
-        # this implicitly also confirms that instructions() returns a newline-separated
-        # value of type 'str'
-        instructions = self.harvester_class.instructions_list()
-        message = (
-            "Most recipes contain more than one instruction, but this recipe test "
-            "did not.  Please check the implementation (and source HTML) and either "
-            "fix the code, or override this method in your test module if you are sure "
-            "this is the expected behaviour."
-        )
-        self.assertGreater(len(instructions), 1, message)
+        self.assertEqual(sorted(actual.ingredients()), sorted(grouped))
+
+    return test_func
+
+
+def load_tests(
+    loader: unittest.TestLoader, standard_tests: unittest.TestSuite, pattern: str
+) -> unittest.TestSuite:
+    """
+    Customise the loading of tests. This function is automatically picked up by the
+    unittest test loader.
+
+    This function dynamically generates the class definition for RecipeTestCase by adding
+    a test function for each pair of testhtml and testjson files found in the
+    tests/test_data directory.
+
+    This also includes the library tests from the tests/library folder as well.
+
+
+    Parameters
+    ----------
+    loader : unittest.TestLoader
+        The instance of TestLoader loading the tests when unittest is run
+    standard_tests : unittest.TestSuite
+        The tests found by loader by loading the tests from the tests module.
+        This is empty and unused.
+    pattern : str
+        Pattern used to identify tests to load.
+        This is unused.
+
+    Returns
+    -------
+    unittest.TestSuite
+        A TestSuite object populated with tests from the pairs of testhtml and testjson
+        files, and the library tests.
+    """
+    test_dir = pathlib.Path("tests/test_data")
+    for host in test_dir.iterdir():
+        if not host.is_dir():
+            continue
+
+        for testhtml in host.glob("*.testhtml"):
+            testjson = testhtml.with_suffix(".json")
+            if not testjson.is_file():
+                continue
+
+            # Add a new function to RecipeTestCase class to test this scraper
+            # The name of this function is derived from the host directory and
+            # testhtml file name.
+            setattr(
+                RecipeTestCase,
+                f"test_{host.stem}_{testhtml.stem}",
```

Comment:
I'm wondering whether it'd be useful and/or acceptable to break the naming convention here so that the emitted test name is not a valid Python identifier, and could include relevant file path(s).

```
======================================================================
FAIL: test:en.wikibooks.org/wikicookbook.testhtml (tests.RecipeTestCase.test:en.wikibooks.org/wikicookbook.testhtml)
----------------------------------------------------------------------
```

Possibly a crazy idea.  Figured I'd mention it, though.

Reply:
It's in interesting idea. We could also change that line to

```suggestion
                str(testhtml),
```
then the error message would show the path to the file.

```
======================================================================
FAIL: tests/test_data/101cookbooks.com/onehundredonecookbooks.testhtml (tests.RecipeTestCase.tests/test_data/101cookbooks.com/onehundredonecookbooks.testhtml)
----------------------------------------------------------------------
```

Reply:
Seems reasonable - the `RecipeTestCase` class is in there in case the developer wants to figure out where the code is, and the failure path is copy-pastable to investigate further.

Maybe the `testjson` file would be more useful to link to, since the developer may need to edit the JSON, but shouldn't need to edit the HTML?

Reply:
Ek, not `testjson` any more -- but hopefully you know what I mean :)